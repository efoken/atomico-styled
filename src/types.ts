import { Interpolation } from "@emotion/serialize";
import { Atomico } from "atomico/types/dom";

export interface Theme {}

export interface StyledOptions {
    label?: string;
    resolver?: (props: object & { theme: Theme }) => any[];
}

type KebabCase<S> = S extends `${infer C}${infer T}`
    ? T extends Uncapitalize<T>
        ? `${Uncapitalize<C>}${KebabCase<T>}`
        : `${Uncapitalize<C>}-${KebabCase<T>}`
    : S;

type KebabCaseProps<Props extends object> = {
    [K in keyof Props as KebabCase<K>]: Props[K];
};

interface StyledComponent<
    Props extends object,
    Base extends HTMLElement = HTMLElement
> extends Atomico<KebabCaseProps<Props>, Base> {}

export interface StyledFunction<
    Props extends object,
    Base extends HTMLElement = HTMLElement
> {
    <AdditionalProps extends object = {}>(
        ...styles: Interpolation<Props & AdditionalProps & { theme: Theme }>[]
    ): StyledComponent<Props & AdditionalProps, Base>;

    (
        template: TemplateStringsArray,
        ...styles: Interpolation<Props & { theme: Theme }>[]
    ): StyledComponent<Props, Base>;

    <AdditionalProps extends object>(
        template: TemplateStringsArray,
        ...styles: Interpolation<Props & AdditionalProps & { theme: Theme }>[]
    ): StyledComponent<Props & AdditionalProps, Base>;
}
