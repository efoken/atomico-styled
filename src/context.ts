import { createContext, useContext } from "atomico/core";
import { Theme } from "./types";

export const ThemeProvider = createContext<Theme>({} as any);

customElements.define("styled-theme-provider", ThemeProvider);

export function useTheme() {
    return useContext(ThemeProvider);
}
