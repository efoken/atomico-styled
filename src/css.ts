import { serializeStyles } from "@emotion/serialize";
import { css as createCss, Sheet } from "atomico/core";
import { compile, middleware, prefixer, serialize, stringify } from "stylis";

const cache: Record<string, string> = {};

export function css(strings: TemplateStringsArray, ...values: any[]): Sheet;

export function css(...values: any[]): Sheet;

export function css(this: any, ...values: any[]) {
    const serialized = serializeStyles(values, cache, this);
    return createCss([
        serialize(
            compile(serialized.styles),
            middleware([prefixer, stringify])
        ),
    ] as any);
}
