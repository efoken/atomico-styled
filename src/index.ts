export * from "./context";
export * from "./css";
export * from "./styled";
export * from "./types";
