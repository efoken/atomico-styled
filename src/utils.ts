export function interleave(values: any[]) {
    const strings = values[0];
    const finalArray = [strings[0]];
    for (let i = 1, { length } = values; i < length; i += 1) {
        finalArray.push(values[i]);
        if (strings[i] != null) {
            finalArray.push(strings[i]);
        }
    }
    return finalArray;
}
