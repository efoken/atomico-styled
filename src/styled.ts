import { useAttributes } from "@atomico/hooks/use-attributes";
import { useCssLightDom } from "@atomico/hooks/use-css-light-dom";
import hash from "@emotion/hash";
import { c, Component, h } from "atomico/core";
import { css } from "./css";
import { CreateStyledComponent } from "./types";
import { interleave } from "./utils";

const elements: Partial<
    Record<keyof HTMLElementTagNameMap, typeof HTMLElement>
> = {
    a: HTMLAnchorElement,
    area: HTMLAreaElement,
    audio: HTMLAudioElement,
    base: HTMLBaseElement,
    blockquote: HTMLQuoteElement,
    body: HTMLBodyElement,
    br: HTMLBRElement,
    button: HTMLButtonElement,
    canvas: HTMLCanvasElement,
    caption: HTMLTableCaptionElement,
    col: HTMLTableColElement,
    colgroup: HTMLTableColElement,
    data: HTMLDataElement,
    datalist: HTMLDataListElement,
    del: HTMLModElement,
    details: HTMLDetailsElement,
    dialog: HTMLDialogElement,
    div: HTMLDivElement,
    dl: HTMLDListElement,
    embed: HTMLEmbedElement,
    fieldset: HTMLFieldSetElement,
    form: HTMLFormElement,
    h1: HTMLHeadingElement,
    h2: HTMLHeadingElement,
    h3: HTMLHeadingElement,
    h4: HTMLHeadingElement,
    h5: HTMLHeadingElement,
    h6: HTMLHeadingElement,
    head: HTMLHeadElement,
    hr: HTMLHRElement,
    html: HTMLHtmlElement,
    iframe: HTMLIFrameElement,
    img: HTMLImageElement,
    input: HTMLInputElement,
    ins: HTMLModElement,
    label: HTMLLabelElement,
    legend: HTMLLegendElement,
    li: HTMLLIElement,
    link: HTMLLinkElement,
    map: HTMLMapElement,
    menu: HTMLMenuElement,
    meta: HTMLMetaElement,
    meter: HTMLMeterElement,
    object: HTMLObjectElement,
    ol: HTMLOListElement,
    optgroup: HTMLOptGroupElement,
    option: HTMLOptionElement,
    output: HTMLOutputElement,
    p: HTMLParagraphElement,
    picture: HTMLPictureElement,
    pre: HTMLPreElement,
    progress: HTMLProgressElement,
    q: HTMLQuoteElement,
    script: HTMLScriptElement,
    select: HTMLSelectElement,
    slot: HTMLSlotElement,
    source: HTMLSourceElement,
    span: HTMLSpanElement,
    style: HTMLStyleElement,
    table: HTMLTableElement,
    tbody: HTMLTableSectionElement,
    td: HTMLTableCellElement,
    template: HTMLTemplateElement,
    textarea: HTMLTextAreaElement,
    tfoot: HTMLTableSectionElement,
    th: HTMLTableCellElement,
    thead: HTMLTableSectionElement,
    time: HTMLTimeElement,
    title: HTMLTitleElement,
    tr: HTMLTableRowElement,
    track: HTMLTrackElement,
    ul: HTMLUListElement,
    video: HTMLVideoElement,
};

export function styled<
    Tag extends keyof HTMLElementTagNameMap,
    Props extends object = {}
>(element: Tag): CreateStyledComponent<Props> {
    return (...rawStyles: any[] | any) => {
        const name = `css-${hash(JSON.stringify(rawStyles))}`;

        const styles =
            rawStyles[0] == null || rawStyles[0].raw == null
                ? rawStyles
                : interleave(rawStyles);

        const Styled: Component = () => {
            const props = useAttributes();
            useCssLightDom(css.apply(props, [":host{", ...styles, "}"]));
            return h("host", null);
        };

        Styled.props = {};

        if (customElements.get(name) == null) {
            customElements.define(name, c(Styled, elements[element]), {
                extends: element,
            });
        }

        return customElements.get(name) as any;
    };
}
