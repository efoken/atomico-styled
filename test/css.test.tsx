import { fixture } from "atomico/test-dom";
import { css } from "../src";

describe("css", () => {
    test("float property", () => {
        const cls1 = css`
            :host {
                float: left;
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("handles more than 10 dynamic properties", () => {
        const cls1 = css`
            :host {
                text-decoration: ${"underline"};
                border-right: solid blue 54px;
                background: ${"white"};
                color: ${"black"};
                display: ${"block"};
                border-radius: ${"3px"};
                padding: ${"25px"};
                width: ${"500px"};
                z-index: ${100};
                font-size: ${"18px"};
                text-align: ${"center"};
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("falsy value in nested selector on object", () => {
        const cls1 = css({
            ":host": { ":hover": { display: null, color: "hotpink" } },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("boolean as value", () => {
        const cls1 = css({
            ":host": {
                display: "flex",
                color: false,
                backgroundColor: true,
            },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("auto px", () => {
        const cls1 = css({
            ":host": {
                display: "flex",
                flex: 1,
                fontSize: 10,
                "--custom": 5,
            },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("random interpolation with undefined values", () => {
        const cls2 = css`
            :host {
                ${undefined};
                justify-content: center;
            }
        `;
        const tree = fixture(<style>{cls2}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("random expression", () => {
        const cls2 = css`
            :host {
                font-size: 20px;
                @media (min-width: 420px) {
                    color: blue;
                    ${css`
                        width: 96px;
                        height: 96px;
                    `};
                    line-height: 40px;
                }
                background: green;
            }
        `;
        const tree = fixture(<style>{cls2}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("simple composition", () => {
        const cls1 = css`
            display: flex;
            &:hover {
                color: hotpink;
            }
        `;
        const cls2 = css`
            :host {
                ${cls1};
                justify-content: center;
            }
        `;
        const tree = fixture(<style>{cls2}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("handles objects", () => {
        const cls1 = css({
            ":host": {
                float: "left",
                display: "flex",
                color: `${"blue"}`,
                fontSize: `${"20px"}`,
                height: 50,
                width: 20,
            },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("handles array of objects", () => {
        const cls1 = css([{ ":host": { height: 50, width: 20 } }, null]);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("computed key is only dynamic", () => {
        const cls1 = css({ ":host": { fontSize: 10, [`w${"idth"}`]: 20 } });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("composition with objects", () => {
        const cls1 = css({
            display: "flex",
            width: 30,
            height: "calc(40vw - 50px)",
            "&:hover": { color: "blue" },
            ":after": {
                content: '" "',
                color: "red",
            },
            "@media(min-width: 420px)": {
                color: "green",
            },
        });
        const cls2 = css`
            :host {
                ${cls1};
                justify-content: center;
            }
        `;

        const tree = fixture(<style>{cls2}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("@supports", () => {
        const cls1 = css`
            :host {
                @supports (display: grid) {
                    display: grid;
                }
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test.skip("nested at rules", () => {
        const cls1 = css`
            :host {
                @supports (display: grid) {
                    display: grid;
                    @supports (display: flex) {
                        display: flex;
                    }
                }
                @media (min-width: 420px) {
                    color: pink;
                    @media (max-width: 500px) {
                        color: hotpink;
                    }
                }
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("nested array", () => {
        const cls1 = css({
            ":host": [
                [{ display: "inline" }],
                [{ display: "inline-block" }],
                [
                    { display: "block" },
                    [
                        { display: "flex" },
                        [
                            { display: "table" },
                            { color: "darkorchid" },
                            [
                                {
                                    fontSize: 16,
                                },
                                [
                                    {
                                        "&:after": {
                                            backgroundColor: "aquamarine",
                                        },
                                    },
                                ],
                            ],
                        ],
                    ],
                ],
            ],
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("explicit false", () => {
        const cls1 = css(false);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("array with explicit false", () => {
        const cls1 = css([[{ ":host": { display: "flex" } }], false]);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("array with explicit true", () => {
        const cls1 = css([[{ ":host": { display: "flex" } }], true]);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("nested", () => {
        const cls1 = css`
            :host {
                color: yellow;
                & .some-class {
                    display: flex;
                    & .some-other-class {
                        background-color: hotpink;
                    }
                    @media (max-width: 600px) {
                        background-color: pink;
                    }
                }
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("explicit &", () => {
        const cls1 = css`
            &.another-class {
                display: flex;
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("falsy property value in object", () => {
        const cls1 = css({
            ":host:": { display: "flex", backgroundColor: undefined },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("registered styles as nested selector value in object", () => {
        const cls1 = css({ display: "flex", backgroundColor: "hotpink" });
        const cls2 = css({ ":hover": cls1 });
        const tree = fixture(<style>{cls2}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("composition stuff", () => {
        const cls1 = css({ ":host": { justifyContent: "center" } });
        const cls2 = css([cls1]);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
        const tree2 = fixture(<style>{cls2}</style>);
        expect(tree2).toMatchSnapshot();
    });

    test("null rule", () => {
        const cls1 = css();
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("css variables", () => {
        const cls1 = css`
            :host {
                --some-var: 1px;
                width: var(--some-var);
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("null value", () => {
        const cls1 = css(null);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("weakmap", () => {
        const styles = { ":host": { display: "flex" } };
        const cls1 = css(styles);
        const cls2 = css(styles);
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
        const tree2 = fixture(<style>{cls2}</style>);
        expect(tree2).toMatchSnapshot();
    });

    test("multiline declaration", () => {
        const cls1 = css`
            display: grid;
            grid:
                "AppBar" auto
                "Main" 1fr
                / 1fr;
        `;

        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("multiline selector", () => {
        const cls1 = css`
            .my-class:hover .its-child {
                background: pink;
            }
        `;

        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("rule after media query", () => {
        const cls1 = css`
            @media (min-width: 600px) {
                :host {
                    color: green;
                }
            }
            :host {
                color: hotpink;
            }
        `;
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("nested at rule", () => {
        const cls1 = css({
            ":host": {
                "@media (min-width: 980px)": {
                    backgroundColor: "blue",
                    "@supports (width: 100vw)": {
                        backgroundColor: "red",
                    },
                },
            },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });

    test("array fallback", () => {
        const cls1 = css({
            ":host": {
                color: ["yellow", "hotpink"],
            },
        });
        const tree = fixture(<style>{cls1}</style>);
        expect(tree).toMatchSnapshot();
    });
});
