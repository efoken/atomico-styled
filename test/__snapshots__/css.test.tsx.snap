// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css @supports 1`] = `
<style>
  Mark {}
  @supports (display: grid){:host{display:-ms-grid;display:grid;}}
  Mark {}
</style>
`;

exports[`css array fallback 1`] = `
<style>
  Mark {}
  :host{color:yellow;color:hotpink;}
  Mark {}
</style>
`;

exports[`css array with explicit false 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css array with explicit true 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css auto px 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex:1;-ms-flex:1;flex:1;font-size:10px;--custom:5;}
  Mark {}
</style>
`;

exports[`css boolean as value 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css composition stuff 1`] = `
<style>
  Mark {}
  :host{-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}
  Mark {}
</style>
`;

exports[`css composition stuff 2`] = `
<style>
  Mark {}
  :host{-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}
  Mark {}
</style>
`;

exports[`css composition with objects 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;width:30px;height:calc(40vw - 50px);-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}:host :hover{color:blue;}:host :after{content:" ";color:red;}@media(min-width: 420px){:host{color:green;}}
  Mark {}
</style>
`;

exports[`css computed key is only dynamic 1`] = `
<style>
  Mark {}
  :host{font-size:10px;width:20px;}
  Mark {}
</style>
`;

exports[`css css variables 1`] = `
<style>
  Mark {}
  :host{--some-var:1px;width:var(--some-var);}
  Mark {}
</style>
`;

exports[`css explicit & 1`] = `
<style>
  Mark {}
  .another-class{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css explicit false 1`] = `
<style>
  Mark {}
  
  Mark {}
</style>
`;

exports[`css falsy property value in object 1`] = `
<style>
  Mark {}
  :host:{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css falsy value in nested selector on object 1`] = `
<style>
  Mark {}
  :host :hover{color:hotpink;}
  Mark {}
</style>
`;

exports[`css float property 1`] = `
<style>
  Mark {}
  :host{float:left;}
  Mark {}
</style>
`;

exports[`css handles array of objects 1`] = `
<style>
  Mark {}
  :host{height:50px;width:20px;}
  Mark {}
</style>
`;

exports[`css handles more than 10 dynamic properties 1`] = `
<style>
  Mark {}
  :host{-webkit-text-decoration:underline;text-decoration:underline;border-right:solid blue 54px;background:white;color:black;display:block;border-radius:3px;padding:25px;width:500px;z-index:100;font-size:18px;text-align:center;}
  Mark {}
</style>
`;

exports[`css handles objects 1`] = `
<style>
  Mark {}
  :host{float:left;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;color:blue;font-size:20px;height:50px;width:20px;}
  Mark {}
</style>
`;

exports[`css multiline declaration 1`] = `
<style>
  Mark {}
  display:-ms-grid;display:grid;grid:"AppBar" auto "Main" 1fr/1fr;
  Mark {}
</style>
`;

exports[`css multiline selector 1`] = `
<style>
  Mark {}
  .my-class:hover .its-child{background:pink;}
  Mark {}
</style>
`;

exports[`css nested 1`] = `
<style>
  Mark {}
  :host{color:yellow;}:host .some-class{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}:host .some-class .some-other-class{background-color:hotpink;}@media (max-width: 600px){:host .some-class{background-color:pink;}}
  Mark {}
</style>
`;

exports[`css nested array 1`] = `
<style>
  Mark {}
  :host{display:inline;display:inline-block;display:block;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;display:table;color:darkorchid;font-size:16px;}:host:after{background-color:aquamarine;}
  Mark {}
</style>
`;

exports[`css nested at rule 1`] = `
<style>
  Mark {}
  @media (min-width: 980px){:host{background-color:blue;}@supports (width: 100vw){:host{background-color:red;}}}
  Mark {}
</style>
`;

exports[`css null rule 1`] = `
<style>
  Mark {}
  
  Mark {}
</style>
`;

exports[`css null value 1`] = `
<style>
  Mark {}
  
  Mark {}
</style>
`;

exports[`css random expression 1`] = `
<style>
  Mark {}
  :host{font-size:20px;background:green;}@media (min-width: 420px){:host{color:blue;width:96px;height:96px;line-height:40px;}}
  Mark {}
</style>
`;

exports[`css random interpolation with undefined values 1`] = `
<style>
  Mark {}
  :host{-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}
  Mark {}
</style>
`;

exports[`css registered styles as nested selector value in object 1`] = `
<style>
  Mark {}
  :hover:display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;background-color:hotpink;
  Mark {}
</style>
`;

exports[`css rule after media query 1`] = `
<style>
  Mark {}
  @media (min-width: 600px){:host{color:green;}}:host{color:hotpink;}
  Mark {}
</style>
`;

exports[`css simple composition 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}:host :hover{color:hotpink;}
  Mark {}
</style>
`;

exports[`css weakmap 1`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;

exports[`css weakmap 2`] = `
<style>
  Mark {}
  :host{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  Mark {}
</style>
`;
